vehicle-simulator/main.go

```diff
@@ -0,0 +1,186 @@
+// Copyright (c) 2023 Contributors to the Eclipse Foundation
+//
+// See the NOTICE file(s) distributed with this work for additional
+// information regarding copyright ownership.
+//
+// This program and the accompanying materials are made available under the
+// terms of the Eclipse Public License 2.0 which is available at
+// https://www.eclipse.org/legal/epl-2.0, or the Apache License, Version 2.0
+// which is available at https://www.apache.org/licenses/LICENSE-2.0.
+//
+// SPDX-License-Identifier: EPL-2.0 OR Apache-2.0
+
+package main
+
+import (
+	"encoding/json"
+	"flag"
+	"fmt"
+	"log/slog"
+	"os"
+	"os/signal"
+	"syscall"
+
+	dittoModel "github.com/eclipse/ditto-clients-golang/model"
+	"github.com/eclipse/ditto-clients-golang/protocol"
+	"github.com/eclipse/ditto-clients-golang/protocol/things"
+	mqtt "github.com/eclipse/paho.mqtt.golang"
+)
+
+type DeviceData struct {
+	DeviceID string `json:"deviceID"`
+	TenantID string `json:"tenantID"`
+	PolicyID string `json:"policyID"`
+}
+
+var mqttClient mqtt.Client
+var deviceData = DeviceData{}
+var telemetrySenderStarted bool
+var selectedConnectorDeviceID string // holds the last deviceID from the previous active connector
+var ctdSubscriptionTopics []string
+
+var stopTelemetrySender = make(chan struct{})
+var controlTelemetrySender = make(chan string)
+
+// Define flag variables
+var ctdEffectDelay = flag.Int("ctd-effect-delay", 30, "In seconds, Configure trunk/alarms state time")
+var periodicUpdatesDelay = flag.Int("periodic-updates-delay", 5, "In seconds, Configure periodic updates (minimum 5)")
+var idleStateTime = flag.Int("idle-state-time", 120, "In seconds, Configure when the system will enter idle state. Idle state will send telemetry data infrequently. Receiving CTD command will reset the idle state.")
+var initialPeriodicUpdatesDelay = 0 // set the initial periodic updates delay in case of idle state active
+
+func main() {
+	flag.Parse()
+	validateFlags()
+
+	keepAlive := make(chan os.Signal, 1)
+	signal.Notify(keepAlive, os.Interrupt, syscall.SIGTERM)
+
+	initializeClient()
+	getDeviceData()
+
+	defer func() { stopTelemetrySender <- struct{}{} }()
```

Comment:
```suggestion
	defer func() {
		if telemetrySenderStarted {
			stopTelemetrySender <- struct{}{}
		}
	}()
```
If telemetry is not started(e.g. suite-connector is not running or not connected to cloud), the app hangs on exit.